salt/orch/force-removal.sls

```diff
@@ -0,0 +1,130 @@
+# could fail, as we don't know the current state of the cluster at this point. The idea is to
+# try to clean as much as possible where possible, but in the end the only thing we can really
+# ensure is that the salt-key of the target is rejected.
+
+# must provide the node (id) to be removed in the 'target' pillar
+{%- set target = salt['pillar.get']('target') %}
+{%- set target_nodename = salt.saltutil.runner('mine.get', tgt=target, fun='nodename')[target] %}
+# services that will be called when trying to unregister as clean as possible
+{%- set services_to_stop = { "kubelet":   ["stop", "remove-pre-reboot"],
+                             "haproxy":   ["remove-pre-reboot"],
+                             "etc-hosts": ["remove-pre-reboot"] } %}
+# nodes that are responding
+{%- set all_up_nodes = salt.saltutil.runner('manage.up', tgt='P@roles:(etcd|(kube-*))', expr_form='compound') %}
+{%- set up_nodes_except_target = salt.saltutil.runner('manage.up', tgt='P@roles:(etcd|(kube-*)) and not ' + target, expr_form='compound') %}
+{%- set etcd_members = salt.saltutil.runner('mine.get', tgt='G@roles:etcd', fun='network.interfaces', tgt_type='compound').keys() %}
+
+# Ensure we mark all nodes with the "as node is being removed" grain.
+# This will ensure the update-etc-hosts orchestration is not run.
+set-cluster-wide-removal-grain:
+  salt.function:
+    - tgt: {{ all_up_nodes|join(',') }}
+    - tgt_type: list
+    - fail_minions: {{ all_up_nodes|join(',') }}
+    - name: grains.setval
+    - arg:
+      - force_removal_in_progress
+      - true
+
+# Try to sync some information before going on. We might use modules that have not been
+# discovered yet on the minions.
+sync-all:
+  salt.function:
+    - tgt: {{ all_up_nodes|join(',') }}
+    - tgt_type: list
+    - fail_minions: {{ all_up_nodes|join(',') }}
+    - names:
+      - saltutil.refresh_pillar
+      - saltutil.refresh_grains
+      - mine.update
+      - saltutil.sync_all
+
+{%- if target in all_up_nodes %}
+# Try to unregister the machine from cluster services and perform minimal housekeeping. It's
+# important to keep every sls in different toplevel stages, so each one will be tried even
+# if previous fails for whatever reason.
+{%- for service, sls_list in services_to_stop.items() %}
+  {%- for sls in sls_list %}
+stop-{{ service }}-{{ sls }}-stage:
+  salt.state:
+    - tgt: {{ target }}
+    - fail_minions: {{ target }}
+    - sls:
+        - {{ service }}.{{ sls }}
+  {%- endfor %}
+{%- endfor %}
+
+# Try to shutdown the target
+shutdown-target:
+  salt.function:
+    - tgt: {{ target }}
+    - name: cmd.run
+    - fail_minions: {{ target }}
+    - arg:
+      - systemctl poweroff
+    - kwarg:
+        bg: True
+{%- endif %}
+
+{%- if target in etcd_members %}
+# Try to remove the node from the etcd cluster from any other working machine.
+{%- set previous_node = none %}
+{%- for node in up_nodes_except_target %}
```

Comment:
If the `super-master` salt-minion is up but this fails there, we just give up? I think we can do better, try from all the machines that can run this. As long as one works, the rest will be omitted (because the `onfail`).